### ISSUE TITLE
TypeError: Cannot read properties of undefined (reading 'length')

### ISSUE THREAD

**Issue by marcj** [wessberg/ts-clone-node#6]
When using this library to clone a node I get following error:

```
TypeError: Cannot read properties of undefined (reading 'length')
    at Object.escapeLeadingUnderscores (/Users/marc/bude/deepkit-framework/packages/type/node_modules/typescript/lib/typescript.js:12575:28)
    at createBaseIdentifier (/Users/marc/bude/deepkit-framework/packages/type/node_modules/typescript/lib/typescript.js:22259:35)
    at Object.createIdentifier (/Users/marc/bude/deepkit-framework/packages/type/node_modules/typescript/lib/typescript.js:22271:24)
    at cloneIdentifier (/Users/marc/bude/deepkit-framework/packages/type/node_modules/ts-clone-node/dist/cjs/index.js:29:28)
    at executeCloneNode (/Users/marc/bude/deepkit-framework/packages/type/node_modules/ts-clone-node/dist/cjs/index.js:1709:16)
    at nextNode (/Users/marc/bude/deepkit-framework/packages/type/node_modules/ts-clone-node/dist/cjs/index.js:1652:19)
    at Object.nextNode (/Users/marc/bude/deepkit-framework/packages/type/node_modules/ts-clone-node/dist/cjs/index.js:1643:98)
    at cloneArrowFunction (/Users/marc/bude/deepkit-framework/packages/type/node_modules/ts-clone-node/dist/cjs/index.js:434:507)
    at executeCloneNode (/Users/marc/bude/deepkit-framework/packages/type/node_modules/ts-clone-node/dist/cjs/index.js:1936:16)
    at nextNode (/Users/marc/bude/deepkit-framework/packages/type/node_modules/ts-clone-node/dist/cjs/index.js:1652:19)
```

The node in question:

```
Node {
  pos: -1,
  end: -1,
  kind: 213,
  id: 0,
  flags: 8,
  modifierFlagsCache: 0,
  transformFlags: 512,
  parent: undefined,
  original: undefined,
  decorators: undefined,
  modifiers: undefined,
  symbol: undefined,
  localSymbol: undefined,
  locals: undefined,
  nextContainer: undefined,
  name: undefined,
  typeParameters: undefined,
  parameters: [ pos: -1, end: -1, hasTrailingComma: false, transformFlags: 0 ],
  type: undefined,
  body: Identifier {
    pos: -1,
    end: -1,
    kind: 79,
    id: 0,
    flags: 8,
    transformFlags: 0,
    parent: undefined,
    original: undefined,
    flowNode: undefined,
    originalKeywordKind: undefined,
    escapedText: '___Ωs'
  },
  equalsGreaterThanToken: Token {
    pos: -1,
    end: -1,
    kind: 38,
    id: 0,
    flags: 8,
    transformFlags: 0,
    parent: undefined
  }
}
```

The `Identifier` in body was created via regular `NodeFactory.createIdentifier`.

Versions:

```
Typescript: 4.5.2
NodeJs: v16.13.0
ts-clone-node: 0.3.29
```

**Comment by marcj** [comment]
It seems clone-identifier uses `node.text` https://github.com/wessberg/ts-clone-node/blob/master/src/clone-node/clone-identifier.ts#L5

which does not exist (anymore?). It should use `unescapeLeadingUnderscores(node.escapedText)` instead.

My workaround is a hook:


```typescript
    const cloneHook = <T extends Node>(node: T, payload: { depth: number }): CloneNodeHook<T> | undefined => {
            if (isIdentifier(node)) {
                //ts-clone-node wants to read `node.text` which does not exist. we hook into into an provide the correct value.
                return {
                    text: () => {
                        return unescapeLeadingUnderscores(node.escapedText);
                    }
                } as any;
            }
            return;
        };
```

**Comment by wessberg** [comment]
Hi there.

`text` is a property of `Identifier` AST nodes, and you'll get this value back when you call `getText()` on such a node. I'm not sure why it's not present on the node you retrieved from a call to `createIdentifier`, and I'm also not actually able to reproduce it, as `text` is not a nullable nor an optional key of the `Identifier` interface. But it may not be an enumerable key, so I'm thinking if you're maybe doing something with the created identifier somewhere like spreading it into a new object literal or something like that? Did your call to `NodeFactory.createIdentifier` look like this: `createIdentifier("___Ωs")`? I'd love to see a minimal repro so that I can understand under which circumstances `text` is undefined

**Comment by marcj** [comment]
There are two possible Identifier objects returned by `createIdentifier` depending on what created the NodeFactory. The one you expect is
https://github.com/microsoft/TypeScript/blob/b2d2f085e249b5e1fdf375dd0c1b182b2c0b8c92/src/services/services.ts#L379
defined in this ObjectAllocator https://github.com/microsoft/TypeScript/blob/b2d2f085e249b5e1fdf375dd0c1b182b2c0b8c92/src/services/services.ts#L872

it has indeed a readonly `text` property: 
```typescript
class IdentifierObject
    ...
    get text(): string {
        return idText(this);
    }
}
```
however this is for services, the compiler has its own ObjectAllocator defined here [getIdentifierConstructor whi](https://github.com/microsoft/TypeScript/blob/ae582a22ee1bb052e19b7c1bc4cac60509b574e0/src/compiler/utilities.ts#L5878) which returns a simpler `Identifier` object https://github.com/microsoft/TypeScript/blob/ae582a22ee1bb052e19b7c1bc4cac60509b574e0/src/compiler/utilities.ts#L5856.

```typescript

    function Identifier(this: Mutable<Node>, kind: SyntaxKind, pos: number, end: number) {
        this.pos = pos;
        this.end = end;
        this.kind = kind;
        this.id = 0;
        this.flags = NodeFlags.None;
        this.transformFlags = TransformFlags.None;
        this.parent = undefined!;
        this.original = undefined;
        this.flowNode = undefined;
    }
```

it does not have any `text` getter. And this is also the interface you get when you type guard for identifier using [`isIdentifier`](https://github.com/microsoft/TypeScript/blob/ec114b8931bf1ad15fd017ecbacf09e099338820/src/compiler/factory/nodeTests.ts#L92) which returns https://github.com/microsoft/TypeScript/blob/ae582a22ee1bb052e19b7c1bc4cac60509b574e0/src/compiler/types.ts#L1124

```
    export interface Identifier extends PrimaryExpression, Declaration {
        readonly kind: SyntaxKind.Identifier;
        /**
         * Prefer to use `id.unescapedText`. (Note: This is available only in services, not internally to the TypeScript compiler.)
         * Text of identifier, but if the identifier begins with two underscores, this will begin with three.
         */
        readonly escapedText: __String;
        readonly originalKeywordKind?: SyntaxKind;                // Original syntaxKind which get set so that we can report an error later
        /*@internal*/ readonly autoGenerateFlags?: GeneratedIdentifierFlags; // Specifies whether to auto-generate the text for an identifier.
        /*@internal*/ readonly autoGenerateId?: number;           // Ensures unique generated identifiers get unique names, but clones get the same name.
        /*@internal*/ generatedImportReference?: ImportSpecifier; // Reference to the generated import specifier this identifier refers to
        isInJSDocNamespace?: boolean;                             // if the node is a member in a JSDoc namespace
        /*@internal*/ typeArguments?: NodeArray<TypeNode | TypeParameterDeclaration>; // Only defined on synthesized nodes. Though not syntactically valid, used in emitting diagnostics, quickinfo, and signature help.
        /*@internal*/ jsdocDotPos?: number;                       // Identifier occurs in JSDoc-style generic: Id.<T>
    }
```

it does not have the `text` property.

I think the problem is that https://github.com/microsoft/TypeScript/blob/64b8172f0688c7567f4a880754a7f972423f092a/src/services/types.ts#L29 is picked up assuming its always available, when it is in fact not.

**Comment by marcj** [comment]
Regarding why you can't reproduce it: the public facing API (when you import 'typescript') uses the `IdentifierObject` object while the `tsc` compiler uses `Identifier`. I use the tsc compiler directly and hook into it my transformer (which uses this library), hence why I get simpler `Identifier` objects only. I think assuming the simpler `Identifier` object and ignoring the simple `text` getter is more robust, at least in this type of scenario. I understand if you don't want to support that though.

**Comment by wessberg** [comment]
> Regarding why you can't reproduce it: the public facing API (when you import 'typescript') uses the `IdentifierObject` object while the `tsc` compiler uses `Identifier`

I use `ts-clone-node` with the Compiler APIs constantly, as it serves as the backing for some of my libraries that expose Custom Transformers that can be passed in the emit phase of a TS Program, for example, as well as for `rollup-plugin-ts`, which constructs a BuilderProgram. So I'm very used to working directly with the Compiler APIs and nodes constructed via the NodeFactory passed to a TypeScript `TransformationContext`. It just surprises me, as in the years I've built tooling around TypeScripts Compiler APIs, this is the first time I ever experience a situation where there can be multiple variants of `Identifier` nodes.

I'll definitely introduce support for this with a check inside the implementation with a workaround similar to your own. I'll mark this as a bug!

**Comment by marcj** [comment]
Oh I'm sorry if I mixed the terminology a bit up. When I was referring to "compiler" I mean the CLI compiler, the "tsc" script. It uses internally the light `Identifier`. The Compiler API (the stuff we get when we import `typescript`) is using the full `IdentifierObject`. When I was saying I hooked into the tsc compiler I actually mean the "tsc" script, not the compiler API.

I you want a PR, I can provide one, but I doubt I get a unit test to work as we don't get the ObjectAllocator of the tsc compiler in the runtime compiler API.
